src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -610,9 +612,7 @@ void Chttp2ServerListener::ActiveConnection::SendGoAway() {
 }
 
 void Chttp2ServerListener::ActiveConnection::Start(
-    RefCountedPtr<Chttp2ServerListener> listener,
     OrphanablePtr<grpc_endpoint> endpoint, const ChannelArgs& args) {
-  listener_ = std::move(listener);
   RefCountedPtr<HandshakingState> handshaking_state_ref;
   {
     MutexLock lock(&mu_);
```

Comment:
The more I look at this code, the more I hate the locking story that we wound up with here. :/  I had to go through the following exercise to convince myself that there weren't any other edge cases here, and I don't see any, but I'm writing down my reasoning, just to provide a sanity check.

It looks like we're doing the following to start handshaking:
- `ActiveConnection::Start()` acquires the lock to grab a ref to the handshaker, then releases the lock before calling `HandshakingState::Start()`.
- `HandshakingState::Start()` re-acquires the lock to grab a ref to the handshake manager, then releases the lock before starting the handshake.

And it looks like we're doing the following to orphan:
- `ActiveConnection::Orphan()` acquires the lock to set `handshaking_state_` to null, then releases the lock before orphaning the `HandshakingState` object.
- `HandshakingState::Orphan()` re-acquires the lock to shutdown the handshake manager.

This means that any time we have thread A starting a connection at the same time as thread B is orphaning it, there are 5 different orders these accesses can occur in, as follows.

In this order, there isn't really any race; the orphan happens after handshaking is started, as you'd expect:
1. `ActiveConnection::Start()`
2. `HandshakingState::Start()`
3. `ActiveConnection::Orphan()`
4. `HandshakingState::Orphan()`

In this order, we'll wind up shutting down the handshake manager before we start the handshake, which I believe is handled properly inside the handshake manager (it will cause the handshake to fail immediately):
1. `ActiveConnection::Start()`
2. `ActiveConnection::Orphan()`
3. `HandshakingState::Start()`
4. `HandshakingState::Orphan()`

This order is the same as the previous one -- the handshake manager will be shut down before handshaking starts, so the handshake will fail immediately:
1. `ActiveConnection::Start()`
2. `ActiveConnection::Orphan()`
3. `HandshakingState::Orphan()`
4. `HandshakingState::Start()`

In this order, there isn't really any race; the orphan happens before handshaking has started, so `ActiveConnection::Start()` will notice that `shutdown_` has been set, and it will be a no-op:
1. `ActiveConnection::Orphan()`
2. `ActiveConnection::Start()`
3. `HandshakingState::Start()`
4. `HandshakingState::Orphan()`

This order is the same as the previous one; there isn't really any race here:
1. `ActiveConnection::Orphan()`
2. `HandshakingState::Orphan()`
3. `ActiveConnection::Start()`
4. `HandshakingState::Start()`

Assuming I haven't missed anything here, I think this looks good.